hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/internal/EntityInitializerImpl.java

```diff
@@ -664,6 +665,14 @@ public void resolveInstance(Object instance, EntityInitializerData data) {
 					data.concreteDescriptor.getIdentifier( data.getInstance(), session )
 			);
 			data.entityHolder = session.getPersistenceContextInternal().getEntityHolder( data.entityKey );
+			if ( data.entityHolder == null ) {
+				// Entity was most probably removed in the same session without setting the reference to null
+				resolveKey( data );
+				assert data.getState() == State.MISSING;
+				assert referencedModelPart instanceof ToOneAttributeMapping
+						&& ( (ToOneAttributeMapping) referencedModelPart ).getSideNature() == TARGET;
+				return;
+			}
```

Comment:
Had to re-introduce the `null` check from https://github.com/hibernate/hibernate-orm/pull/8506 since when the new setting is disabled this can happen.